components/core/src/clp_s/ArchiveReader.cpp

```diff
@@ -3,81 +3,187 @@
 #include "ReaderUtils.hpp"
 
 namespace clp_s {
-void ArchiveReader::open(ArchiveReaderOption& option) {
-    // Open dictionary readers
-    m_archive_path = option.archive_path;
+void ArchiveReader::open(std::string const& archive_path) {
+    if (m_is_open) {
+        throw OperationFailed(ErrorCodeNotReady, __FILENAME__, __LINE__);
+    }
+    m_is_open = true;
+    m_archive_path = archive_path;
 
     m_var_dict = ReaderUtils::get_variable_dictionary_reader(m_archive_path);
     m_log_dict = ReaderUtils::get_log_type_dictionary_reader(m_archive_path);
     m_array_dict = ReaderUtils::get_array_dictionary_reader(m_archive_path);
 
-    m_var_dict->read_new_entries();
-    m_log_dict->read_new_entries();
-    m_array_dict->read_new_entries();
+    m_table_file_reader.open(m_archive_path + "/table");
+    m_metadata_file_reader.open(m_archive_path + "/metadata");
+}
+
+void ArchiveReader::read_metadata() {
+    constexpr size_t cDecompressorFileReadBufferCapacity = 64 * 1024;  // 16 KB
+    m_metadata_decompressor.open(m_metadata_file_reader, cDecompressorFileReadBufferCapacity);
```

Comment:
Since the pattern for all of these readX functions is basically
```
m_X_decompressor.open(X_file_reader)
// do reading
m_X_decompressor.close()
```
do we really need a ZstdDecompressor for every different read function?
Could we instead create a single `m_archive_decompressor` and reuse it for all of these different functions? It would probably require some changes to the ZstdDecompressor class so that it can re-use the same buffer each time it gets reopened, but it might be worth it.

Curious what you think? If you don't think its worth it I won't push for it.

Reply:
Yeah one decompressor object that we use for both, and we change the ZstdDecompressor implementation so that it can keep its internal buffer when it gets closed.

Reply:
For the first case, I don't think it can reuse the buffer because it needs to read from a different file. For the second case, the buffer can be reused in two different ways. 1): create a memory mapped file and use `CompressedDataBuf` as `InputType` (what glt does) 2): create a method `Decompressor::try_seek_to_file_pos`. In this method, we call `FileReader::try_seek_to_pos` and reset the `m_compressed_stream_block`, but still use the same `m_file_read_buffer`. A potential optimization is to record `current_file_pos` after each  and see if we can use some data from `m_compressed_stream_block`. Which one do you prefer? And do you think I should add it in this PR or another PR?

Reply:
Potentially the first, but yeah we can leave it for another PR and talk about it more later.